src/pmem.c

```diff
@@ -602,17 +653,243 @@ memmove_nodrain_normal(void *pmemdest, const void *src, size_t len)
+		d += 8;
+	}
+
+	/* Copy the tail (<128 bytes) in 16 bytes chunks */
+	len &= CHUNK_MASK;
+	if (len != 0) {
+		cnt = len >> MOVNT_SHIFT;
+		for (i = 0; i < cnt; i++) {
+			xmm0 = _mm_loadu_si128(s);
+			_mm_stream_si128(d, xmm0);
+			s++;
+			d++;
+		}
+	}
+
+	/* Copy the last bytes (<16). First dwords then bytes */
+	len &= MOVNT_MASK;
+	if (len != 0) {
+		cnt = len >> DWORD_SHIFT;
+		int32_t	*d32 = (int32_t *)d;
+		int32_t	*s32 = (int32_t *)s;
+
+		for (i = 0; i < cnt; i++) {
+			_mm_stream_si32(d32, *s32);
+			d32++;
+			s32++;
+		}
+
+		cnt =	len & DWORD_MASK;
+		uint8_t	*d8 = (uint8_t *)d32;
+		const uint8_t	*s8 = (uint8_t *)s32;
+
+		for (i = 0; i < cnt; i++) {
+			*d8 = *s8;
+			d8++;
+			s8++;
+		}
+		pmem_flush(d8, cnt);
+	}
+	return tempdest;
+}
+
 /*
  * memmove_nodrain_movnt -- (internal) memmove to pmem without hw drain, movnt
  */
 static void *
 memmove_nodrain_movnt(void *pmemdest, const void *src, size_t len)
 {
+
+	int	i;
+	uint64_t chunk_size;
+	uint64_t num_chunks;
+	size_t cnt = 0;
+	void *dest1 = pmemdest;
+	void *src1 = (void *)src;
+	int mod = 0;
+
 	/* way too chatty for LOG level 3 */
 	LOG(15, "pmemdest %p src %p len %zu", pmemdest, src, len);
 
-	/* XXX stub version */
-	return memmove(pmemdest, src, len);
+	/*
+	 * The way this works is:
+	 * 	Calculate the non-overlapping size
+	 * 	Break the memmove into these chunk sizes
+	 *	Determine if we are copying forward/backward
+	 * This handles all cases, including overlap. Consider the following:
+	 *
+	 * 	src
+	 * high addr				dest
+	 *					high addr
+	 * ------
+	 * |	|s3-----------------------------------------------V
+	 * ------						--------
+	 * |	|s2-----------------------------|		|	|d3
+	 * ------				|-------|	--------
+	 * |	|s1---------------------------- 	|------>|	|d2
+	 * ------				|		--------
+	 *					|-------------->|	|d1
+	 * **** 						--------
+	 *
+	 * In this case we copy down, starting with s1 to d1, so that we
+	 * never overwrite the source before moving it. Any leftover
+	 * bytes are handled after moving the chunk size data.
+	 * This works similarly when src < dest.
+	 */
+	if (src > pmemdest)
+		chunk_size = util_nonoverlap_range((void *)src1, dest1, len);
+	else
+		chunk_size = util_nonoverlap_range(dest1, src1, len);
+
+	/*
+	 * The most common case is non-overlapping addresses.
+	 */
+	if (chunk_size == len) {
+		pmem_align(&dest1, src1, &cnt, UP);
+		src1 += cnt;
+		len -= cnt;
+		return (pmemdest = (nodrain_movnt(dest1, src1, len)));
+	}
+
+	/* overlapping addresses, possible unaligned dest. */
+	if ((uint64_t)src1 > (uint64_t)dest1) {
+		pmem_align(&dest1, src1, &cnt, UP);
+		len -= cnt;
+		src1 += cnt;
+
+		/* Align the chunk size */
+		if (chunk_size > MOVNT_SIZE) {
+			if ((mod = (chunk_size % MOVNT_SIZE)) != 0)
+				chunk_size -= mod;
+		}
+		num_chunks = len/chunk_size;
+
+		for (i = 0; i < num_chunks; i ++) {
+			nodrain_movnt(dest1, src1, chunk_size);
+			src1 += chunk_size;
+			len -= chunk_size;
```

Comment:
Not to be too picky, but this line can be moved out of the loop:
```C
len -= (num_chunks * chunk_size);
```
or (if you prefer), it can be done w/o *num_chunks* variable:
```C
while (len >= chunk_size) {
    nodrain_movnt(dest1, src1, chunk_size);
    src1 += chunk_size;
    dest1 += chunk_size;
    len -= chunk_size;
}
```